### PR TITLE
routerrpc: check payaddr before using for probing

### PR DESCRIPTION
**Pull request by ellemouton**
As in other places where we extract and use the payment address from an invoice, we first check that it is set before dereferencing it.

### CODE REVIEW

File: docs/release-notes/release-notes-0.19.0.md [hunk 1]
```diff
@@ -23,6 +23,9 @@
   propagate mission control and debug level config values to the main LND config
   struct so that the GetDebugInfo response is accurate.
 
+* [Fix a bug]() that would cause a nil pointer dereference during the probing
```
Comment (Roasbeef) [inline docs/release-notes/release-notes-0.19.0.md:26]: Can update to `9134`.

File: routing/pathfind_test.go [hunk 2]
```diff
@@ -3502,11 +3502,11 @@ func TestLastHopPayloadSize(t *testing.T) {
 			t.Parallel()
 
 			var mpp *record.MPP
-			if tc.restrictions.PaymentAddr != nil {
-				mpp = record.NewMPP(
-					tc.amount, *tc.restrictions.PaymentAddr,
-				)
-			}
+			tc.restrictions.PaymentAddr.WhenSome(
```
Comment (Roasbeef) [inline routing/pathfind_test.go:3505]: I think could also be: 
```go
mpp := fn.MapOptionZ(func(p [32]byte) *record.MPP {
    return record.NewMPP(tc.amount, addr)
})
```

File: routing/router.go [hunk 3]
```diff
@@ -1063,9 +1063,13 @@ func (r *ChannelRouter) PreparePayment(payment *LightningPayment) (
 	switch {
 	// If this is an AMP payment, we'll use the AMP shard tracker.
 	case payment.amp != nil:
+		var addr [32]byte
+		payment.PaymentAddr.WhenSome(func(i [32]byte) {
```
Comment (Roasbeef) [inline routing/router.go:1067]: Can be: 
```go
addr := payment.PaymentAddr.UnwrapOr([32]byte{})
```

File: rpcserver.go [hunk 4]
```diff
@@ -7381,9 +7382,9 @@ func (r *rpcServer) DecodePayReq(ctx context.Context,
 
 	// Extract the payment address from the payment request, if present.
 	var paymentAddr []byte
-	if payReq.PaymentAddr != nil {
-		paymentAddr = payReq.PaymentAddr[:]
-	}
+	payReq.PaymentAddr.WhenSome(func(addr [32]byte) {
```
Comment (Roasbeef) [inline rpcserver.go:7385]:
```go
paymentAddr := payReq.PaymentAddr.UnwrapOr([32]byte{})
```

File: zpay32/encode.go [hunk 5]
```diff
@@ -301,13 +301,12 @@ func writeTaggedFields(bufferBase32 *bytes.Buffer, invoice *Invoice) error {
...
 	}
-	if invoice.PaymentAddr != nil {
-		err := writeBytes32(
-			bufferBase32, fieldTypeS, *invoice.PaymentAddr,
-		)
-		if err != nil {
-			return err
-		}
+	var err error
+	invoice.PaymentAddr.WhenSome(func(addr [32]byte) {
```
Comment (Roasbeef) [inline zpay32/encode.go:305]:
```go
err := fm.MapOptionZ(func(addr [32]byte) {
    return writeBytes32(bufferBase32, fieldTypeS, addr)
})
```

**Review by Roasbeef — CHANGES_REQUESTED** [review]

File: routing/pathfind_test.go (hunk 2 above)
Comment (ellemouton) [reply routing/pathfind_test.go:3505]: oooo cool - TIL

File: routing/router.go (hunk 3 above)
Comment (ellemouton) [reply routing/router.go:1067]: niiiice

File: invoices/modification_interceptor.go [hunk 6]
```diff
@@ -163,11 +167,19 @@ func (s *HtlcModificationInterceptor) RegisterInterceptor(
 
 // Start starts the service.
 func (s *HtlcModificationInterceptor) Start() error {
+	if !s.started.CompareAndSwap(false, true) {
+		return nil
```
Comment (yyforyongyu) [inline invoices/modification_interceptor.go:171]: nit: add debug log like others?

File: server.go [hunk 7]
```diff
@@ -2411,6 +2411,10 @@ func (s *server) Stop() error {
 		if err := s.invoices.Stop(); err != nil {
 			srvrLog.Warnf("failed to stop invoices: %v", err)
 		}
+		if err := s.interceptableSwitch.Stop(); err != nil {
```
Comment (yyforyongyu) [inline server.go:2414]: 🙏

**Review by yyforyongyu — APPROVED** [review]
Thanks for the quick fix!

File: routing/pathfind_test.go (hunk 2 above)
Comment (Roasbeef) [reply routing/pathfind_test.go]: Yeah the diff with this one is that if the option is `None`, then it'll return the zero value of the return type. So in this case, it'll just return `nil` if the option isn't set.

**Review by Roasbeef — APPROVED** [review]
LGTM 🎊

**Pull request merged**